// eslint-plugin-validate-filename/dist/rules/limit-extensions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.limitExtensions = void 0;
const path_1 = __importDefault(require("path"));
const micromatch_1 = __importDefault(require("micromatch"));
const meta = {
    type: 'layout',
    docs: {
        description: 'Only certain extensions are allowed.',
        recommended: false,
        url: '',
    },
    messages: {
        errorNoMatchExtensionType: '{{filename}} does not match filename extension type. Only {{extensions}} is allowed.',
        errorNoMatchExtensionTypes: '{{filename}} does not match filename extension type. Only {{extensions}} are allowed.',
    },
    schema: [
        {
            type: 'object',
            additionalProperties: false,
            properties: {
                rules: {
                    type: 'array',
                    items: {
                        type: 'object',
                        additionalProperties: false,
                        properties: {
                            target: {
                                type: 'string',
                            },
                            extensions: {
                                type: 'array',
                                items: {
                                    type: 'string',
                                },
                            },
                        },
                    },
                },
            },
        },
    ],
};
exports.limitExtensions = {
    meta,
    create: (context) => {
        return {
            Program: (node) => {
                const { rules } = context.options[0];
                const filePath = context.filename;
                if (!rules || !filePath) {
                    return;
                }
                /**
                 * check target filename
                 **/
                const targetRule = rules.find((v) => micromatch_1.default.isMatch(filePath, v.target));
                if (targetRule) {
                    const filename = path_1.default.basename(filePath);
                    const ext = path_1.default.extname(filePath);
                    if (!targetRule.extensions.includes(ext)) {
                        context.report({
                            node,
                            messageId: targetRule.extensions.length === 1
                                ? 'errorNoMatchExtensionType'
                                : 'errorNoMatchExtensionTypes',
                            data: {
                                filename,
                                extensions: targetRule.extensions.join(', '),
                            },
                        });
                        return;
                    }
                }
            },
        };
    },
};
