// eslint-plugin-validate-filename/dist/rules/naming-rules.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.namingRules = void 0;
const path_1 = __importDefault(require("path"));
const micromatch_1 = __importDefault(require("micromatch"));
const utils_1 = require("../utils");
const meta = {
    type: 'layout',
    docs: {
        description: '',
        category: '',
        recommended: false,
        url: '',
    },
    messages: {
        errorNoMatchCase: '{{file}} does not match filename case. Rename it to {{caseType}} case.',
        errorNoMatchPattern: '{{file}} does not match filename pattern.',
    },
    schema: [
        {
            type: 'object',
            additionalProperties: false,
            properties: {
                rules: {
                    type: 'array',
                    items: {
                        type: 'object',
                        additionalProperties: false,
                        properties: {
                            case: {
                                enum: Object.keys(utils_1.regexCaseMap),
                                type: 'string',
                            },
                            target: {
                                type: 'string',
                            },
                            patterns: {
                                type: 'string',
                            },
                            excludes: {
                                type: 'array',
                                items: {
                                    type: 'string',
                                },
                            },
                        },
                        required: ['target'],
                        anyOf: [
                            {
                                required: ['case'],
                            },
                            {
                                required: ['patterns'],
                            },
                        ],
                    },
                },
            },
        },
    ],
};
exports.namingRules = {
    meta,
    create: (context) => {
        return {
            Program: (node) => {
                const { rules } = context.options[0];
                const filePath = context.filename;
                if (!rules || !filePath) {
                    return;
                }
                /**
                 * target filename
                 **/
                const dirPath = path_1.default.dirname(filePath);
                const dirName = path_1.default.basename(dirPath);
                const targetRule = rules.find((v) => {
                    var _a;
                    return micromatch_1.default.isMatch(filePath, v.target) &&
                        !((_a = v.excludes) === null || _a === void 0 ? void 0 : _a.includes(dirName));
                });
                if (targetRule) {
                    const ext = path_1.default.extname(filePath);
                    const file = path_1.default.basename(filePath);
                    const filename = (0, utils_1.getFilename)(path_1.default.basename(filePath, ext));
                    if (filename === 'index') {
                        return;
                    }
                    if (targetRule.case) {
                        if (!micromatch_1.default.isMatch(
                        // Consider cases with leading underscores. (ex: _document.tsx)
                        filename.replace(/^_/, ''), utils_1.regexCaseMap[targetRule.case])) {
                            context.report({
                                node,
                                messageId: 'errorNoMatchCase',
                                data: {
                                    file,
                                    caseType: targetRule.case,
                                },
                            });
                            return;
                        }
                    }
                    if (targetRule.patterns) {
                        const pattern = new RegExp(targetRule.patterns);
                        if (!pattern.test(file)) {
                            context.report({
                                node,
                                messageId: 'errorNoMatchPattern',
                                data: {
                                    file,
                                },
                            });
                            return;
                        }
                    }
                }
            },
        };
    },
};
